#################### Prerequisite #################### 
# Follow the information below to download latest gapseq:
# https://github.com/jotech/gapseq
git clone https://github.com/jotech/gapseq

# Follow the information below to set up an environment (depends on your working system) to run gapseq:
# https://gapseq.readthedocs.io/en/latest/install.html#centos-fedora-rhel

# For example, for Centos/Fedora/RHEL:

module load R/3.6.1 					# Load this module if you are about to run gapseq on UNSW HPC katana

sudo yum install ncbi-blast+ git glpk-devel BEDTools exonerate hmmer bc
git clone https://github.com/tseemann/barrnap.git
export PATH=$PATH:barrnap/bin/barrnap # needs to be permanent => .bashrc ?
R -e 'install.packages(c("data.table", "stringr", "sybil", "getopt", "reshape2", "doParallel", "foreach", "R.utils", "stringi", "glpkAPI", "CHNOSZ", "jsonlite"))'
R -e 'if (!requireNamespace("BiocManager", quietly = TRUE)) install.packages("BiocManager"); BiocManager::install("Biostrings")'
git clone https://github.com/jotech/gapseq && cd gapseq


#################### Run gapseq #################### 
# Quickstart testing is recommended before analysing your own data: https://github.com/jotech/gapseq

# Load dependencies on Katana, skip this if you work on your own computer.
module load perl/5.28.0 			# Load this module if you are about to run gapseq on UNSW HPC katana
module load git/2.22.0 				# Load this module if you are about to run gapseq on UNSW HPC katana
module load bedtools/2.27.1 		# Load this module if you are about to run gapseq on UNSW HPC katana
module load blast+/2.9.0 			# Load this module if you are about to run gapseq on UNSW HPC katana
module load hmmer/3.2.1 			# Load this module if you are about to run gapseq on UNSW HPC katana
module load glpk/4.65 				# Load this module if you are about to run gapseq on UNSW HPC katana
module load barrnap/0.9 			# Load this module if you are about to run gapseq on UNSW HPC katana
module load gcc/7.3.0 				# Load this module if you are about to run gapseq on UNSW HPC katana
module load exonerate/2.2.0 		# Load this module if you are about to run gapseq on UNSW HPC katana
module load parallel/20190522 		# Load this module if you are about to run gapseq on UNSW HPC katana
module load libsbml/5.18.0 			# Load this module if you are about to run gapseq on UNSW HPC katana
module add R/3.6.1  				# Load this module if you are about to run gapseq on UNSW HPC katana
module load cplex/12.9.0-academic   # Load this module if you are about to run gapseq on UNSW HPC katana



# step 1: Search for metabolic pathways of your genome (e.g. STY_Merged_OTU01.fasta):
cd ./PATH_to_genome/

./PATH_to_gapseq/gapseq find -p all -t Bacteria -b 200 -c 75 -l all -y ./STY_Merged_OTU01.fasta
#√   -p keywords such as pathways or subsystems (for example amino,nucl,cofactor,carbo,polyamine)
#√   -t Taxonomic range for sequences to be downloaded (default: Bacteria)
#√   -b Bit score cutoff for local alignment (default: 200)
#√   -c Coverage cutoff for local alignment (default: 75)
#√   -l Select the pathway database (MetaCyc(2712), KEGG(523), SEED(666), all(3922); default: metacyc,custom)


# step 2: Search for transporter of your genome (e.g. STY_Merged_OTU01.fasta):
./PATH_to_gapseq/gapseq find-transport -b 50 -c 70 ./STY_Merged_OTU01.fasta
#√   -b bit score cutoff for local alignment (default: 50)
#√   -c coverage cutoff for local alignment (default: 75)


# step 3: Create a draft model:
./PATH_to_gapseq/gapseq draft -r ./STY_Merged_OTU01-all-Reactions.tbl -t ./STY_Merged_OTU01-Transporter.tbl -p ./STY_Merged_OTU01-all-Pathways.tbl -c ./STY_Merged_OTU01.fasta -u 100 -l 50 -b neg
#√     -r|--blast.res          Blast-results table generated by gapseq.sh.
#√     -t|--transporter.res    Blast-results table generated by transporter.sh.
#√     -b|--biomass            Gram "pos" OR "neg" OR "archae" OR "auto"? Default: "auto". Please note: if set to "auto", the external programms barrnap, usearch, and bedtools are required.
#√     -u|--high.evi.rxn.BS    Reactions with an associated blast-hit with a bitscore above this value will be added to the draft model as core reactions (i.e. high-sequence-evidence reactions)
#√     -l|--min.bs.for.core    Reactions with an associated blast-hit with a bitscore below this value will be considered just as reactions that have no blast hit.



# step 4: Fill the gene/reaction gap based on the growth medium you provide:
# Gapseq include several medium diet as choices, check more details here https://gapseq.readthedocs.io/en/latest/usage/medium.html
./PATH_to_gapseq/gapseq fill -m ./STY_Merged_OTU01-draft.RDS -n ./PATH_to_gapseq/dat/media/meerwasser.csv -c ./STY_Merged_OTU01-rxnWeights.RDS -g ./STY_Merged_OTU01-rxnXgenes.RDS -b 50 -o ./STY_Merged_OTU01_meerwasser -r TRUE

#√     -m|--model                  GapSeq-Draft-Model to be gapfilled (RDS or SBML)
#√     -n|--media                  tab- or komma separated table for media components. Requires three named columns: 1 - "compounds" (for metab. IDs), 2 - "name" (metab. name), 3 - "maxFlux" (maximum inflow flux)
#√     -c|--rxn.weights.file       Reaction weights table generated by gapseq function "generate_GSdraft.R" (RDS format).
#√     -g|--rxnXgene.table         Table with gene-X-reaction associations as generated by the "generate_GSdraft.R" (RDS format)
#√     -b|--bcore                  Minimum bitscore for reaction associated blast hits to consider reactions as core/candidate reactions. Default: 50
#√     -o|--output.dir             Directory to store results. Default: "gapfill".
#√     -r|--relaxed.constraints    Save final model as unconstraint network (i.e. all exchange reactions are open). Default: FALSE
